### PR TITLE
Geomap: Add ability to select a data query filter for each layer

### LINKED ISSUE #48274: Geomap panel only shows the first query
**What happened**:
geomap panel only shows the first datasource query data

**How to reproduce it (as minimally and precisely as possible)**:

using **testdata** plugin to generate the data

<details>
  <summary>dashboard json:</summary>

```json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "datasource": {
        "type": "testdata",
        "uid": "_N83fAQnk"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "rgba(245, 54, 54, 0.9)",
                "value": null
              },
              {
                "color": "rgba(237, 129, 40, 0.89)",
                "value": 0
              },
              {
                "color": "rgba(50, 172, 45, 0.97)",
                "value": 10
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "maxDataPoints": 1,
      "options": {
        "basemap": {
          "name": "Basemap",
          "type": "default"
        },
        "controls": {
          "mouseWheelZoom": false,
          "showAttribution": true,
          "showDebug": false,
          "showScale": false,
          "showZoom": true
        },
        "layers": [
          {
            "config": {
              "showLegend": true,
              "style": {
                "color": {
                  "fixed": "dark-green"
                },
                "opacity": 0.4,
                "rotation": {
                  "fixed": 0,
                  "max": 360,
                  "min": -360,
                  "mode": "mod"
                },
                "size": {
                  "fixed": 5,
                  "max": 15,
                  "min": 2
                },
                "symbol": {
                  "fixed": "img/icons/marker/circle.svg",
                  "mode": "fixed"
                },
                "textConfig": {
                  "fontSize": 12,
                  "offsetX": 0,
                  "offsetY": 0,
                  "textAlign": "center",
                  "textBaseline": "middle"
                }
              }
            },
            "location": {
              "latitude": "x",
              "longitude": "y",
              "mode": "coords"
            },
            "name": "Layer 1",
            "tooltip": true,
            "type": "markers"
          },
          {
            "config": {
              "showLegend": true,
              "style": {
                "color": {
                  "fixed": "dark-green"
                },
                "opacity": 0.4,
                "rotation": {
                  "fixed": 0,
                  "max": 360,
                  "min": -360,
                  "mode": "mod"
                },
                "size": {
                  "fixed": 5,
                  "max": 15,
                  "min": 2
                },
                "symbol": {
                  "fixed": "img/icons/marker/circle.svg",
                  "mode": "fixed"
                },
                "textConfig": {
                  "fontSize": 12,
                  "offsetX": 0,
                  "offsetY": 0,
                  "textAlign": "center",
                  "textBaseline": "middle"
                }
              }
            },
            "location": {
              "latitude": "x2",
              "longitude": "y2",
              "mode": "coords"
            },
            "name": "Layer 2",
            "tooltip": true,
            "type": "markers"
          }
        ],
        "view": {
          "id": "zero",
          "lat": 0,
          "lon": 0,
          "zoom": 1
        }
      },
      "pluginVersion": "8.5.0",
      "targets": [
        {
          "csvContent": "name,x,y\na,1,1\nb,2,2",
          "datasource": {
            "type": "testdata",
            "uid": "_N83fAQnk"
          },
          "refId": "A",
          "scenarioId": "csv_content"
        },
        {
          "csvContent": "name,x2,y2\nc,5,5\nd,6,6",
          "datasource": {
            "type": "testdata",
            "uid": "_N83fAQnk"
          },
          "refId": "B",
          "scenarioId": "csv_content"
        }
      ],
      "title": "Panel Title",
      "type": "geomap"
    }
  ],
  "schemaVersion": 36,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "New dashboard",
  "version": 0,
  "weekStart": ""
}
```
</details>

**Environment**:
- Grafana version: Grafana v8.5.0 (6134e3cf3)

### PR DESCRIPTION
**Pull request by mmandrus**
**What this PR does / why we need it**:
Right now, the Geomap panel always uses the first query in the list, so users are not able to fully visualize all of their data. This fix allows users to select which query to use in each layer. It also includes some handling for when queries are renamed or deleted that will keep the UI up to date.

**Which issue(s) this PR fixes**:
Fixes #48274

### CODE REVIEW

**Comment by mmandrus** [comment]
Added query picker under the layer type 
image

File: packages/grafana-data/src/geo/layer.ts [hunk 1]
```diff
@@ -102,6 +106,12 @@ export interface MapLayerRegistryItem<TConfig = MapLayerOptions> extends Registr
    */
   showOpacity?: boolean;
 
+  /**
+   * Whether this layer type renders its features from a data frame
+   * Required field as this determines key options layout logic
+   */
+  usesDataFrame: boolean;
```
Comment (ryantxu) [inline packages/grafana-data/src/geo/layer.ts:113]:
```suggestion
  usesData: boolean;
```
frames seems a bit specific?  or maybe "usesQueryResult?"

File: packages/grafana-data/src/geo/layer.ts [hunk 2]
```diff
@@ -58,6 +58,9 @@ export interface MapLayerOptions<TConfig = any> {
   // Common method to define geometry fields
   location?: FrameGeometrySource;
 
+  // Defines which data query is associated with the layer
+  dataquery?: string;
```
Comment (ryantxu) [inline packages/grafana-data/src/geo/layer.ts:62]:
```suggestion
  filterData?: MatcherConfig;
```
and refId can be one of the options

**Comment by mmandrus** [comment]
Query rename behavior
dataqueryrename

File: packages/grafana-data/src/geo/layer.ts (hunk 2 above)
Comment (mmandrus) [reply packages/grafana-data/src/geo/layer.ts:62]: I'm not sure what I'm supposed to do here. this field corresponds to the builder path specified [here](https://github.com/grafana/grafana/blob/e1dc2b028f54d1b553fa837030ec8c3321a9d67e/public/app/plugins/panel/geomap/editor/layerEditor.tsx#L88). I haven't been able to find other examples that use Matchers in this way.

**Comment by mmandrus** [comment]
This exhibits some strange behavior when the data source is streaming, as the refId is missing from the dropdown, but otherwise it does the trick.

**Review by ryantxu — APPROVED** [review]
@mmandrus -- this looks great, thanks for sticking with it :)

I pushed a small commit that changes the text displayed in the dropdown.  I think we can improve this component in a new PR.

I recall you had previously implemented logic that looked tried to re-sync the current value with a new value when the refID changes.  I think that is OK, but should happen in `FrameSelectionEditor` not the base layer implementation

**Comment by mmandrus** [comment]
Hate to keep this one going, but while re-adding the query ID recovery logic, I discovered a bug in the onFilterChange callback. If we don't add the layer name as a dependency, we always use the "last" version of the callback, which updates the wrong layer.

Also re-added the recovery logic, using redux to store the previous dropdown state after every update.

**Review by drew08t — APPROVED** [review]
Looks good to me, thanks!

**Pull request merged**